### PR TITLE
[1.1] Socket buffered write. Write full cpack at once

### PR DESCRIPTION
**Pull request by nullkiller**
Already boosted game load significantly. Will take a look onto read as well.

### CODE REVIEW

**Comment by dydzio0614** [comment]
I wonder if there is chance to boost hero movement speed in multiplayer by utilizing this feature

**Comment by IvanSavenko** [comment]
> I wonder if there is chance to boost hero movement speed in multiplayer by utilizing this feature

Don't expect much. There may be some improvement especially on low-end system (read: Android phones), but most of hero movement lag comes from TCP connection. 
Especially when playing over Wi-Fi or mobile networks (3G/4G/etc)

**Comment by nullkiller** [comment]
If somebody see a better solution for this or any caveats please let me know. I hardly understand what I have done.

File: lib/serializer/Connection.h [hunk 1]
```diff
@@ -63,8 +65,16 @@ class DLL_LINKAGE CConnection
...
 	int write(const void * data, unsigned size) override;
 	int read(void * data, unsigned size) override;
+	void flushBuffers();
 
 	std::shared_ptr<boost::asio::io_service> io_service; //can be empty if connection made from socket
+
+	bool enableBufferedWrite;
+	boost::asio::streambuf writeBuffer;
+
+	bool enableBufferedRead;
```
Comment (Nordsoft91) [inline lib/serializer/Connection.h:75]:
```suggestion
	bool enableBufferedWrite = false;
	boost::asio::streambuf writeBuffer;

	bool enableBufferedRead = false;
```

**Review by Nordsoft91 — COMMENTED** [review]
Why not to keep bufferization always enabled?

**Comment by IvanSavenko** [comment]
> Why not to keep bufferization always enabled?

I am not nullkiller, but I would also be cautious.
Currently change is controlled - we know that bufferization is only done during that single call.
If there is some code path that writes data to Connection without generating CPack then enabling bufferization always will stop that code from working (since it won't be sent until next CPack)

**Pull request merged**